source/Ship.cpp

```diff
@@ -86,6 +86,39 @@ namespace {
 			fuel -= transfer * fuelCost;
 		}
 	}
+	
+	// Helper function to reduce a given status effect according 
+	// to its resistance, limited by how much energy and fuel are available.
+	// Updates the stat and the energy, fuel, and heat amounts.
+	void DoStatusEffect(bool isDisabled, double &stat, double resistance, double &energy, double energyCost, double &fuel, double fuelCost, double &heat, double heatCost)
+	{
+		if(isDisabled || resistance <= 0)
+		{
+			stat = .99 * stat;
+			return;
+		}
+		
+		// Calculate how much resistance can be used assuming no
+		// energy or fuel cost.
+		double after = max(0., .99 * stat - resistance);
+		resistance = .99 * stat - after;
+		
+		// Limit the resistance by the available energy and fuel
+		if(energyCost)
+			resistance = min(resistance, energy / energyCost);
+		if(fuelCost)
+			resistance = min(resistance, fuel / fuelCost);
+		
+		// Update the energy, fuel, and heat given how much resistance is being used,
+		// then update the stat.
+		if(resistance > 0.)
+		{
+			energy -= resistance * energyCost;
+			fuel -= resistance * fuelCost;
+			heat += resistance * heatCost;
```

Comment:
We should account for negative heatCost, i.e. being required to supply heat to use the resistance.